File: Day-12/bloom-filter-sample/python/bloom_filter.py
from __future__ import annotations

import hashlib
import math


class BloomFilter:
    def __init__(self, size_bits: int, hash_count: int) -> None:
        if size_bits <= 0:
            raise ValueError("size_bits must be positive")
        if hash_count <= 0:
            raise ValueError("hash_count must be positive")

        self.size_bits = size_bits
        self.hash_count = hash_count
        self._bits = bytearray((size_bits + 7) // 8)
        self._insertions = 0

    def add(self, item: str) -> None:
        for index in self._indexes(item):
            self._set_bit(index)
        self._insertions += 1

    def might_contain(self, item: str) -> bool:
        return all(self._get_bit(index) for index in self._indexes(item))

    def fill_ratio(self) -> float:
        bits_set = sum(byte.bit_count() for byte in self._bits)
        return bits_set / self.size_bits

    def estimated_false_positive_rate(self) -> float:
        m = self.size_bits
        k = self.hash_count
        n = self._insertions
        return (1 - math.exp((-k * n) / m)) ** k

    def _indexes(self, item: str) -> list[int]:
        if not item:
            raise ValueError("item must be non-empty")

        return [self._index_for_seed(item, seed) for seed in range(self.hash_count)]

    def _index_for_seed(self, item: str, seed: int) -> int:
        digest = hashlib.sha256(f"{seed}:{item}".encode("utf-8")).digest()
        return int.from_bytes(digest[:8], byteorder="big") % self.size_bits

    def _set_bit(self, index: int) -> None:
        byte_index, bit_offset = divmod(index, 8)
        self._bits[byte_index] |= 1 << bit_offset

    def _get_bit(self, index: int) -> bool:
        byte_index, bit_offset = divmod(index, 8)
        return bool(self._bits[byte_index] & (1 << bit_offset))

File: Day-12/bloom-filter-sample/java/BloomFilter.java
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.BitSet;

public class BloomFilter {
    private final int sizeBits;
    private final int hashCount;
    private final BitSet bits;
    private int insertions;

    public BloomFilter(int sizeBits, int hashCount) {
        if (sizeBits <= 0) {
            throw new IllegalArgumentException("sizeBits must be positive");
        }
        if (hashCount <= 0) {
            throw new IllegalArgumentException("hashCount must be positive");
        }

        this.sizeBits = sizeBits;
        this.hashCount = hashCount;
        this.bits = new BitSet(sizeBits);
        this.insertions = 0;
    }

    public void add(String item) {
        for (int index : indexes(item)) {
            bits.set(index);
        }
        insertions += 1;
    }

    public boolean mightContain(String item) {
        for (int index : indexes(item)) {
            if (!bits.get(index)) {
                return false;
            }
        }
        return true;
    }

    public double fillRatio() {
        return (double) bits.cardinality() / sizeBits;
    }

    public double estimatedFalsePositiveRate() {
        return Math.pow(1 - Math.exp((-1.0 * hashCount * insertions) / sizeBits), hashCount);
    }

    private int[] indexes(String item) {
        if (item == null || item.isBlank()) {
            throw new IllegalArgumentException("item must be non-empty");
        }

        int[] indexes = new int[hashCount];
        for (int seed = 0; seed < hashCount; seed += 1) {
            indexes[seed] = indexForSeed(item, seed);
        }
        return indexes;
    }

    private int indexForSeed(String item, int seed) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashed = digest.digest((seed + ":" + item).getBytes(StandardCharsets.UTF_8));
            long value = ByteBuffer.wrap(hashed, 0, Long.BYTES).getLong();
            return (int) Math.floorMod(value, sizeBits);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 algorithm unavailable", e);
        }
    }
}
